common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,434 @@
+                decrypterInit = true;
+                ciphertextRead += aesGcmHkdfStreaming.getHeaderLength();
+            }
+            return true;
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object ciphertextMessage)
+                throws GeneralSecurityException {
+            Preconditions.checkArgument(ciphertextMessage instanceof ByteBuf,
+                    "Unrecognized message type: %s",
+                    ciphertextMessage.getClass().getName());
+            ByteBuf ciphertextNettyBuf = (ByteBuf) ciphertextMessage;
+            // The format of the output is:
+            // [8 byte length][Internal IV and header][Ciphertext][Auth Tag]
+            try {
+                if (!initalizeExpectedLength(ciphertextNettyBuf)) {
+                    // We have not read enough bytes to initialize the expected length.
+                    return;
+                }
+                if (!initalizeDecrypter(ciphertextNettyBuf)) {
+                    // We nave not read enough bytes to initalize a header, needed to
+                    // initialize a decrypter.
+                    return;
+                }
+                if (expectedLength == ciphertextRead) {
+                    // If the expected length is just the header, the ciphertext is 0 length.
+                    lastSegment = true;
+                }
+                while (ciphertextNettyBuf.readableBytes() > 0 && !lastSegment) {
+                    ciphertextBuffer.clear();
+                    // Read the ciphertext into the local buffer
+                    int readableBytes = Integer.min(
+                            ciphertextNettyBuf.readableBytes(),
+                            ciphertextBuffer.remaining());
+                    if (readableBytes == 0) {
+                        return;
+                    }
+                    int expectedRemaining = (int) (expectedLength - ciphertextRead);
+                    int bytesToRead = Integer.min(readableBytes, expectedRemaining);
+                    // The smallest ciphertext size is 16 bytes for the auth tag
+                    ciphertextBuffer.limit(bytesToRead);
+                    ciphertextNettyBuf.readBytes(ciphertextBuffer);
+                    ciphertextRead += bytesToRead;
+                    // Check if this is the last segment
+                    if (ciphertextRead == expectedLength) {
+                        lastSegment = true;
+                    } else if (ciphertextRead > expectedLength) {
+                        throw new IllegalStateException("Read more ciphertext than expected.");
+                    }
+                    plaintextBuffer.clear();
+                    ciphertextBuffer.flip();
+
+                    decrypter.decryptSegment(
+                            ciphertextBuffer,
+                            segmentNumber,
+                            lastSegment,
+                            plaintextBuffer);
```

Comment:
We have the analogous problem here as what we fixed in encryption side.
`ciphertextNettyBuf` size need not be aligned to `ciphertextBuffer` capacity - and so reads from it can be incomplete to create a full segment (`ciphertextRead < expectedLength` after read above)
We have to accumulate it in `ciphertextBuffer` and decrypt only when it is the expected size.

Reply:
We should add a test case for this as well.

Reply:
Addressed this and added a test case for it.